File: ports-rb/url_parsing.rb
require "uri"
require_relative "ports"

suite "suites/url-parsing-RFC.ports" do
  placeholder "url-parse" do |_env, url_string|
    URI.parse(url_string)
  rescue URI::InvalidURIError => e
    e
  end

  placeholder "parse-error?" do |_env, parse_result|
    parse_result.is_a?(URI::InvalidURIError)
  end

  placeholder "url-scheme" do |_env, uri|
    uri.scheme
  end

  placeholder "url-authority" do |_env, uri|
    authority = uri.host
    authority = "#{uri.userinfo}@#{authority}" if uri.userinfo
    authority = "#{authority}:#{uri.port}" if uri.port
    authority
  end
end

File: rosetta-test-js/url-parsing.js
import { suite } from './rosetta-test.js';
import { URL } from 'url';

const urlParsingSuite = suite("stdlib url parsing", "rosetta-test-suites/url-parsing-rfc.ros");

urlParsingSuite.placeholder("url-parse", (env, urlString) => {
    try {
        return new URL(urlString);
    } catch (e) {
        return e;
    }
});

urlParsingSuite.placeholder("parse-error?", (env, parseResult) => {
    return parseResult instanceof Error;
});

urlParsingSuite.placeholder("url-scheme", (env, parseResult) => { 
    // WhatWG protocol is the scheme without the colon
    return parseResult instanceof URL ? parseResult.protocol.replace(':', '') : "";
});

urlParsingSuite.placeholder("url-authority", (env, parseResult) => {
    return parseResult instanceof URL ? parseResult.host : "";
});

// Running
urlParsingSuite.run({
    expectedFailures: ["test_ipv6_host"], 
    excludeCapabilities: ["root.authority"]});
